### PR TITLE
4837 single user email import

### PR DESCRIPTION
**Pull request by jomapp**
Draft pull request for single user email import.

Notes:

- [x] Email import documentation (in tutadb) will be updated in the following days.
- [ ] There are several open TODOs to be found in the documentation.
- [ ] Main entry points are: `ImapAdSync.ts` as well as `ImapImporter.ts`.
- [ ] See mail from 06/30/23.

### CODE REVIEW

File: src/gui/base/icons/Icons.ts [hunk 1]
```diff
@@ -137,6 +139,8 @@ export const IconsSvg: Record<Icons, string> = Object.freeze({
...
 	Palette: `<svg viewBox="0 0 512 512" xmlns="http://www.w3.org/2000/svg"><path d="M430.1 347.9c-6.6-6.1-16.3-7.6-24.6-9-11.5-1.9-15.9-4-22.6-10-14.3-12.7-14.3-31.1 0-43.8l30.3-26.9c46.4-41 46.4-108.2 0-149.2-34.2-30.1-80.1-45-127.8-45-55.7 0-113.9 20.3-158.8 60.1-83.5 73.8-83.5 194.7 0 268.5 41.5 36.7 97.5 55 152.9 55.4h1.7c55.4 0 110-17.9 148.8-52.4 14.4-12.7 12-36.6.1-47.7zM120 216c0-17.7 14.3-32 32-32s32 14.3 32 32-14.3 32-32 32-32-14.3-32-32zm40 126c-17.7 0-32-14.3-32-32s14.3-32 32-32 32 14.3 32 32-14.3 32-32 32zm64-161c-17.7 0-32-14.3-32-32s14.3-32 32-32 32 14.3 32 32-14.3 32-32 32zm72 219c-26.5 0-48-21.5-48-48s21.5-48 48-48 48 21.5 48 48-21.5 48-48 48zm24-208c-17.7 0-32-14.3-32-32s14.3-32 32-32 32 14.3 32 32-14.3 32-32 32z"/></svg>`,
+	Play: `<svg viewBox="0 0 512 512" height="512" width="512" xmlns="http://www.w3.org/2000/svg"><path d="M128 96v320l256-160L128 96z"/></svg>`,
```
Comment (charlag) [inline src/gui/base/icons/Icons.ts:142]: I don't know if we really need icons for these things

File: src/settings/imapimport/AddImapImportWizard.ts [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+	constructor(readonly config: ImapImportModelConfig) {
+		this._imapAccountHost = stream(config.imapAccountHost)
+		this._imapAccountPort = stream(config.imapAccountPort)
+		this._imapAccountUsername = stream(config.imapAccountUsername)
+		this._imapAccountPassword = stream(config.imapAccountPassword)
+		this._rootImportMailFolderName = stream(config.rootImportMailFolderName)
+		this._imapImportState = new ImapImportState(ImportState.NOT_INITIALIZED)
+	}
+
+	get imapAccountHost(): Stream<string> {
```
Comment (charlag) [inline src/settings/imapimport/AddImapImportWizard.ts:44]: I don't think it makes sense to re-wrap all of them in getters/setters, also I would prefer to avoid streams unless you have a strong reason to use them

File: src/settings/imapimport/AddImapImportWizard.ts [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+	readonly imapAccountPassword: string
+	readonly rootImportMailFolderName: string
+}
+
+export class ImapImportModel {
+	private _imapAccountHost: stream<string>
+	private _imapAccountPort: stream<string>
+	private _imapAccountUsername: stream<string>
+	private _imapAccountPassword: stream<string>
+	private revealImapAccountPassword: boolean = false
```
Comment (charlag) [inline src/settings/imapimport/AddImapImportWizard.ts:31]: does this have to be in the model? it is kinda view state

File: src/settings/imapimport/ConfigureImapImportPage.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+import { assertMainOrNode } from "../../api/common/Env"
+import { InitializeImapImportParams } from "../../api/worker/imapimport/ImapImporter.js"
+import { showProgressDialog } from "../../gui/dialogs/ProgressDialog.js"
+import { locator } from "../../api/main/MainLocator.js"
+import { ImapImportState, ImportState } from "../../api/worker/imapimport/ImapImportState.js"
+import { AddImapImportData } from "./AddImapImportWizard.js"
+
+assertMainOrNode()
+
+const DEFAULT_IMAP_IMPORT_MAX_QUOTA = "2500000000"
```
Comment (charlag) [inline src/settings/imapimport/ConfigureImapImportPage.ts:17]: what does it mean?

File: src/settings/imapimport/EnterImapCredentialsPage.ts [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+		this.dom = vnode.dom as HTMLElement
+	}
+
+	view(vnode: Vnode<WizardPageAttrs<AddImapImportData>>): Children {
+		const imapAccountHostAttrs: TextFieldAttrs = {
+			label: "imapAccountHost_label",
+			value: vnode.attrs.data.model.imapAccountHost(),
+			oninput: vnode.attrs.data.model.imapAccountHost,
+			helpLabel: () => {
+				const host = vnode.attrs.data.model.imapAccountHost()
```
Comment (charlag) [inline src/settings/imapimport/EnterImapCredentialsPage.ts:29]: what is host for here?

File: src/settings/imapimport/EnterImapCredentialsPage.ts [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+				const errorMsg = vnode.attrs.data.model.validateImapAccountHost()
+
+				if (errorMsg) {
+					return lang.get(errorMsg)
+				} else {
+					return lang.get("imapAccountHost_helpLabel")
+				}
+			},
+		}
+		const imapAccountPortAttrs: TextFieldAttrs = {
```
Comment (charlag) [inline src/settings/imapimport/EnterImapCredentialsPage.ts:39]: should this be a number field?

File: src/settings/imapimport/ImapImportSettingsViewer.ts [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+	}
+
+	private async pauseImapImport() {
+		let newImapImportState = await locator.imapImporterFacade.pauseImport()
+		this.imapImportState(newImapImportState)
+	}
+
+	private async requestImapImportAccountSyncState() {
+		const importImapAccountSyncState = await locator.imapImporterFacade.loadImportImapAccountSyncState()
+		const rootImportMailFolder = await locator.imapImporterFacade.loadRootImportFolder()
```
Comment (charlag) [inline src/settings/imapimport/ImapImportSettingsViewer.ts:191]: can it fail? what happens?

File: src/settings/imapimport/ImapImportSettingsViewer.ts [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+	private rootImportMailFolderName: string = ""
+	private imapAccountHost: Stream<string>
+	private imapAccountPort: Stream<string>
+	private imapAccountUsername: Stream<string>
+	private stars: Stream<string>
+	private importedMailCount: stream<string>
+	private imapImportState: stream<ImapImportState>
+
+	constructor() {
+		this.imapAccountHost = stream<string>("")
```
Comment (charlag) [inline src/settings/imapimport/ImapImportSettingsViewer.ts:33]: steams are also unnecessary here

File: tsconfig.json [hunk 9]
```diff
@@ -7,7 +7,7 @@
 		"outDir": "build/dist",
 		"incremental": true
 	},
-	"include": ["src/", "libs/*.ts", "types/*.d.ts"],
+	"include": ["src/", "libs/*.ts", "types/*.d.ts", "node_modules/imapflow/lib/types.d.ts"],
```
Comment (charlag) [inline tsconfig.json:10]: other types are picked up automatically, why is this necessary?;

**Review by charlag — COMMENTED** [review]
Great job! It's a complicated machine with a lot of parts and I think we need to really make it as understandable as possible.

I've left some comments in places but it would be too much to repeat it every time.

Here are some general code style comments:
 * use for-of loop instead of foreach whenever possible
 * use const instead of let whenever possible
 * collapse fields and constructors whenever possible
 * prefer plain objects/interfaces to classes, especially in remote invocation context
 * do not use streams unless you really need them, prefer plain fields
 * do not prepare view attributes before rendering, just use them whenever you render a component

Besides that we need a lot of JSDOC comments for different methods/classes. I know there's another doc being prepared and it's good to have high-level overview there but it should still be possible to grasp what the class does without referring to it.

**Review by charlag — COMMENTED** [review]
Great job! It's a complicated machine with a lot of parts and I think we need to really make it as understandable as possible.

I've left some comments in places but it would be too much to repeat it every time.

Here are some general code style comments:
 * use for-of loop instead of foreach whenever possible
 * use const instead of let whenever possible
 * collapse fields and constructors whenever possible
 * prefer plain objects/interfaces to classes, especially in remote invocation context
 * do not use streams unless you really need them, prefer plain fields
 * do not prepare view attributes before rendering, just use them whenever you render a component

Besides that we need a lot of JSDOC comments for different methods/classes. I know there's another doc being prepared and it's good to have high-level overview there but it should still be possible to grasp what the class does without referring to it.

File: src/desktop/imapimport/adsync/utils/AdSyncUtils.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+export type TimeIntervalTimeStamp = string
+
+export type Throughput = number // in bytes per ms
+export type AverageThroughput = number // in bytes per ms
+
+export type AverageEfficiencyScore = number
+
+export type DownloadBatchSize = number
+
+export function getAverageOfList(list: number[]) {
```
Comment (ganthern) [inline src/desktop/imapimport/adsync/utils/AdSyncUtils.ts:11]: would this have a place in arrayUtils

File: src/desktop/imapimport/adsync/utils/DifferentialUidLoader.ts [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+
+		let toSeq = fromSeq + downloadBatchSize
+		let isFinalBatch = totalMessageCount == null || toSeq > totalMessageCount
+
+		let fromSeqString = fromSeq.toString()
+		let toSeqString = totalMessageCount == null ? "*" : isFinalBatch ? totalMessageCount.toString() : toSeq.toString()
+
+		let mails = this.imapClient.fetch(`${fromSeqString}:${toSeqString}`, { uid: true }, {})
+
+		// @ts-ignore
```
Comment (ganthern) [inline src/desktop/imapimport/adsync/utils/DifferentialUidLoader.ts:93]: huh

File: src/desktop/imapimport/adsync/utils/DifferentialUidLoader.ts [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+	async getNextUidFetchRequest(downloadBatchSize: number): Promise<UidFetchRequest | null> {
+		let waitingUidFetchRequest = this.uidFetchRequestQueue.pop()
+		if (waitingUidFetchRequest) {
+			return waitingUidFetchRequest
+		}
+
+		let nextUidFetchRange: number[]
+		let fetchRequestType: UidFetchRequestType
+
+		if (this.emitAdSyncEventTypes.has(AdSyncEventType.CREATE) && this.uidCreateQueue.length != 0) {
```
Comment (ganthern) [inline src/desktop/imapimport/adsync/utils/DifferentialUidLoader.ts:121]: I think this are flags that are populated from the config to turn on/off certain events

File: src/desktop/imapimport/adsync/utils/DifferentialUidLoader.ts [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+				fetchRequestType: fetchRequestType,
+				usedDownloadBatchSize: downloadBatchSize,
+			}
+		})
+
+		this.uidFetchRequestQueue.push(...nextUidFetchSequenceRequests)
+		return this.uidFetchRequestQueue.pop() ?? null
+	}
+
+	private buildUidFetchSequenceStrings(uidsToFetch: number[]): string[] {
```
Comment (ganthern) [inline src/desktop/imapimport/adsync/utils/DifferentialUidLoader.ts:149]: this needs to be a freestanding function with tests

File: src/desktop/imapimport/adsync/ImapAdSync.ts [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+	processesTimeToLive: number
+	isEnableDownloadBatchSizeOptimizer: boolean
+	downloadBatchSizeOptimizationDifference: number
+	defaultDownloadBatchSize: number
+	optimizationInterval: number
+	emitAdSyncEventTypes: Set<AdSyncEventType>
+	isEnableImapQresync: boolean
+}
+
+export class ImapAdSync {
```
Comment (ganthern) [inline src/desktop/imapimport/adsync/ImapAdSync.ts:29]: I think this could be combined with the session? it's not really doing anything

File: src/desktop/imapimport/adsync/ImapSyncSession.ts [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+
+		imapMailbox.subFolders?.forEach((imapMailbox) => {
+			result.push(...this.traverseImapMailboxes(knownMailboxes, imapMailbox))
+		})
+		return result
+	}
+
+	onStartSyncSessionProcess(processId: number, nextMailboxToDownload: ImapSyncSessionMailbox): void {
+		if (this.state == SyncSessionState.RUNNING) {
+			console.log("onStartSyncSessionProcess : processId: " + processId + " -> " + nextMailboxToDownload.mailboxState.path)
```
Comment (ganthern) [inline src/desktop/imapimport/adsync/ImapSyncSession.ts:189]: concat in console.log

File: src/desktop/imapimport/adsync/ImapSyncSessionMailbox.ts [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+	set specialUse(value: ImapMailboxSpecialUse | null) {
+		this._specialUse = value
+
+		switch (this._specialUse) {
+			case ImapMailboxSpecialUse.INBOX:
+				this.importance = SyncSessionMailboxImportance.HIGH
+				break
+			case ImapMailboxSpecialUse.TRASH:
+			case ImapMailboxSpecialUse.ARCHIVE:
+			case ImapMailboxSpecialUse.ALL:
```
Comment (ganthern) [inline src/desktop/imapimport/adsync/ImapSyncSessionMailbox.ts:52]: can we skip this, since it would give us all mails again?

File: src/desktop/imapimport/adsync/ImapSyncSessionProcess.ts [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+			)
+
+			differentialUidLoader
+				.calculateUidDiff(
+					this.adSyncOptimizer.optimizedSyncSessionMailbox.lastFetchedMailSeq,
+					this.adSyncOptimizer.optimizedSyncSessionMailbox.downloadBatchSize,
+					this.adSyncOptimizer.optimizedSyncSessionMailbox.mailCount,
+				)
+				.then((deletedUids) => {
+					this.handleDeletedUids(deletedUids, openedImapMailbox, adSyncEventListener)
```
[the remark this answers is not in the document]
Comment (ganthern) [reply src/desktop/imapimport/adsync/ImapSyncSessionProcess.ts:116]: apparently, it's intentional to prevent the imap server from limiting us because of repetitive commands

File: src/desktop/imapimport/adsync/ImapSyncSessionProcess.ts [hunk 18]
```diff
@@ -0,0 +1,294 @@
...
+					continue
+				}
+
+				this.adSyncOptimizer.optimizedSyncSessionMailbox.reportDownloadBatchSizeUsage(nextUidFetchRequest.usedDownloadBatchSize)
+
+				let mails = imapClient.fetch(
+					nextUidFetchRequest.uidFetchSequenceString,
+					{
+						uid: true,
+						// @ts-ignore
```
Comment (ganthern) [inline src/desktop/imapimport/adsync/ImapSyncSessionProcess.ts:136]: huh

File: src/desktop/imapimport/adsync/ImapSyncSessionProcess.ts [hunk 19]
```diff
@@ -0,0 +1,294 @@
...
+		this.adSyncOptimizer.optimizedSyncSessionMailbox.mailCount = imapMailboxStatus.messageCount ?? null
+	}
+
+	private async handleDeletedUids(deletedUids: number[], openedImapMailbox: ImapMailbox, adSyncEventListener: AdSyncEventListener) {
+		deletedUids.forEach((deletedUid) => {
+			this.emitImapMailDeleteEvent(deletedUid, openedImapMailbox, adSyncEventListener)
+		})
+	}
+
+	private setupImapFlowErrorHandler(imapClient: ImapFlow, adSyncEventListener: AdSyncEventListener) {
```
Comment (ganthern) [inline src/desktop/imapimport/adsync/ImapSyncSessionProcess.ts:271]: should this be made use-once
